docs/doc/90-contributing/09-how-to-write-aggregate-functions.md

```diff
@@ -2,135 +2,126 @@
+- **accumulate** accumulates count and value of `AggregateAvgState` for non-null rows in the block.
+- **accumulate_keys** gets the corresponding state value with `place.get::<AggregateAvgState<SumT>>()`, and updates it.
 
-Let's take into the function `accumulate_keys`, because this is the only function that a little hard to understand in this case.
-
-```sql
+```rust
 fn accumulate_keys(
     &self,
     places: &[StateAddr],
     offset: usize,
-    columns: &[ColumnRef],
+    columns: &[Column],
     _input_rows: usize,
 ) -> Result<()> {
-    if columns[0].data_type().data_type_id() == TypeID::Boolean {
-        // boolean cast into u8 column
-        // ...
-    } else {
-        let darray: &PrimitiveColumn<T> = unsafe { Series::static_cast(&columns[0]) };
-        darray.iter().zip(places.iter()).for_each(|(c, place)| {
-            let place = place.next(offset);
-            let state = place.get::<AggregateSumState<SumT>>();
-            state.add(c.as_());
-        });
-    }
-
+    let darray = NumberType::<T>::try_downcast_column(&columns[0]).unwrap();
+    darray.iter().zip(places.iter()).for_each(|(c, place)| {
+        let place = place.next(offset);
+        let state = place.get::<AggregateAvgState<SumT>>();
+        state.add(c.as_(), 1);
+    });
     Ok(())
 }
 ```
 
-The `places` is the memory address of the first state in this row, so we can get the address of `AggregateSumState<T>` using `places[row] + offset`, then using `place.get::<AggregateSumState<SumT>>()` to get the value of the corresponding state.
-
-Since we already know the array type of this function, we can safely cast it to arrow's `PrimitiveArray<T>`, here we make two branches to reduce the branch prediction of CPU, `null` and `no_null`. In `no_null` case, we just iterate the array and apply the `sum`, this is good for compiler to optimize the codes into vectorized codes.
+## Refer to other examples
 
-Ok, this example is pretty easy. If you already read this, you may have the ability to write a new function.
+As you see, adding a new aggregate function in Databend is not as hard as you think.
+Before you start to add one, please refer to other aggregate function examples, such as `sum`, `count`.
 
+- [sum](https://github.com/datafuselabs/databend/blob/d5e06af03ba0f99afdd6bdc974bf2f5c1c022db8/src/query/functions/src/aggregates/aggregate\_sum.rs)
+- [count](https://github.com/datafuselabs/databend/blob/d5e06af03ba0f99afdd6bdc974bf2f5c1c022db8/src/query/functions/src/aggregates/aggregate\_count.rs)
 
 ## Testing
-To be a good engineer, don't forget to test your codes, please add unit tests and stateless tests after you finish the new aggregate functions.
 
-- [Unit tests](https://github.com/datafuselabs/databend/blob/a0869e8ac3cc70105822e6f0aaddfa534c4270d2/common/functions/tests/it/aggregates/aggregate_function.rs)
+To be a good engineer, don't forget to test your codes, please add unit tests and stateless tests after you finish the new aggregate functions.
```

Comment:
As a good developer, you always test your code, don't you? Please add unit tests and stateless tests after you complete the new aggregate functions.